msm/src/serialization/constraints.rs

```diff
@@ -185,3 +185,71 @@ impl<F: PrimeField> InterpreterEnv<F> for Env<F> {
         }))
     }
 }
+
+impl<Fp: PrimeField> Env<Fp> {
+    #[allow(dead_code)]
+    // FIXME: not mut
+    fn constrain_lookups(&mut self) -> Vec<E<Fp>> {
+        assert_eq!(self.rangecheck4_lookups.len(), 20);
+        assert_eq!(self.rangecheck15_lookups.len(), 17);
+
+        {
+            let rc4_t_lookup = Lookup {
+                table_id: LookupTable::RangeCheck4,
+                numerator: curr_cell(Column::LookupMultiplicity(
+                    LookupTable::RangeCheck4.to_u32(),
+                )),
+                value: vec![curr_cell(Column::LookupFixedTable(
+                    LookupTable::RangeCheck4.to_u32(),
+                ))],
+            };
+            self.rangecheck4_lookups.push(rc4_t_lookup);
+        }
+
+        {
+            let rc15_t_lookup = Lookup {
+                table_id: LookupTable::RangeCheck15,
+                numerator: curr_cell(Column::LookupMultiplicity(
+                    LookupTable::RangeCheck15.to_u32(),
+                )),
+                value: vec![curr_cell(Column::LookupFixedTable(
+                    LookupTable::RangeCheck15.to_u32(),
+                ))],
+            };
+            self.rangecheck15_lookups.push(rc15_t_lookup);
+        }
+
+        // This can be generalized for any table. We can have a hashmap or an
+        // array of lookups
+        // Computing individual "boat"
+        let mut constraints = vec![];
+        let mut idx = 0;
+        for chunk in self.rangecheck4_lookups.chunks(6) {
+            constraints.push(combine_lookups(
+                Column::LookupPartialSum(idx),
+                chunk.to_vec(),
+            ));
+            idx += 1;
```

Comment:
perhaps add an assert of what's the expected index at this point

Reply:
Opening https://github.com/o1-labs/proof-systems/issues/1987